src/transport/SecureSessionTable.h

```diff
@@ -133,21 +91,99 @@ class SecureSessionTable
-        });
-        return result != nullptr ? MakeOptional<SessionHandle>(*result) : Optional<SessionHandle>::Missing();
-    }
+    Optional<SessionHandle> FindSecureSessionByLocalKey(uint16_t localSessionId);
 
 private:
+    friend class TestSecureSessionTable;
+
+    /**
+     * This provides a sortable wrapper for a SecureSession object. A SecureSession
+     * isn't directly sortable since it is not swappable (i.e meet criteria for ValueSwappable).
+     *
+     * However, this wrapper has a stable pointer to a SecureSession while being swappable with
+     * another instance of it.
+     *
+     */
+    struct SortableSession
+    {
+    public:
+        void swap(SortableSession & other)
+        {
+            SortableSession tmp(other);
+            other.mSession = mSession;
+            mSession       = tmp.mSession;
+        }
+
+        const Transport::SecureSession * operator->() const { return mSession; }
+
+    private:
+        SecureSession * mSession;
+        friend class SecureSessionTable;
+    };
+
+    /**
+     *
+     * Encapsulates all the necessary context for an eviction policy callback
+     * to implement its specific policy. The context is provided to the callee
+     * with the expectation that it'll call Sort() with a comparator function provided
+     * to get the list of sessions sorted in the desired order.
+     *
+     */
+    class EvictionPolicyContext
+    {
+    public:
+        /*
+         * Called by the policy implementor to sort the list of sessions given a comparator
+         * function. The provided function shall have the following signature:
+         *
+         * bool CompareFunc(const SortableSession &a, const SortableSession &b);
+         *
+         * If a is a better candidate than b, true should be returned. Else, return false.
+         *
+         * NOTE: Sort() can be called multiple times.
+         *
+         */
+        template <typename CompareFunc>
+        void Sort(CompareFunc func)
+        {
+            std::sort(mSessionList.begin(), mSessionList.begin() + mSessionList.size(), func);
+        }
+
+        const ScopedNodeId & GetSessionEvictionHint() const { return mSessionEvictionHint; }
+
+    private:
+        EvictionPolicyContext(Span<SortableSession> sessionList, ScopedNodeId sessionEvictionHint)
+        {
+            mSessionList         = sessionList;
+            mSessionEvictionHint = sessionEvictionHint;
+        }
+
+        friend class SecureSessionTable;
+        Span<SortableSession> mSessionList;
+        ScopedNodeId mSessionEvictionHint;
+    };
+
+    /**
+     *
+     * This implements the following eviction policy:
+     *
+     *  - Sessions sorted by activity time, with older ones placed ahead of newer ones. This ensures
+     *    we're prioritizing reaping less active sessions over more recently active sessions (activity
+     *    in either TX or RX).
+     *  - Then, sessions sorted by their state - defunct ones ahead of active, ahead
```

Comment:
This makes it sounds like state sort is within same activity time... but it's the other way around: the primary sort key is state, followed by activity time, no?